lupa/_lupa.pyx

```diff
@@ -1609,6 +1642,40 @@ cdef call_lua(LuaRuntime runtime, lua_State *L, tuple args):
     push_lua_arguments(runtime, L, args)
     return execute_lua_call(runtime, L, len(args))
 
+# adapted from https://stackoverflow.com/a/9672205
+cdef void* _lua_alloc_restricted(void* ud, void* ptr, size_t osize, size_t nsize) nogil:
+    cdef size_t* left = <size_t*>ud
+
+    if ptr is NULL:
+        # <http://www.lua.org/manual/5.2/manual.html#lua_Alloc>:
+        # When ptr is NULL, osize encodes the kind of object that Lua is
+        # allocating.
+        # Since we don’t care about that, just mark it as 0.
+        osize = 0
+
+    if nsize == 0:
+        free(ptr)
```

Comment:
We shouldn't use plain C memory allocation. If we do the allocation ourselves, we should either do what Lua does internally, and if that really is plain C allocation, then we should use Python's memory allocator instead.

Reply:
> if that really is plain C allocation

It is. This is the function `luaL_newstate` uses:
```c
     static void *l_alloc (void *ud, void *ptr, size_t osize,
                                                size_t nsize) {
       (void)ud;  (void)osize;  /* not used */
       if (nsize == 0) {
         free(ptr);
         return NULL;
       }
       else
         return realloc(ptr, nsize);
     }
```

Reply:
Python's memory allocator relies on GIL

Reply:
When using python's allocator, I'm getting sporadic segfaults and this: `malloc: *** error for object 0x7fec8551b280: pointer being freed was not allocated`

Reply:
Ok, then ISTM that we need to handle whatever Lua returns us here, which are `malloc()` allocated memory blocks.
(And yes, Python's allocator requires the GIL and does more than just calling `malloc` and friends, potentially much more.)